storage/src/mmr/journaled.rs

```diff
@@ -195,32 +237,65 @@ impl<B: Blob, E: RStorage<B> + Clock + Metrics, H: Hasher> Mmr<B, E, H> {
 
     /// Sync any new elements to disk.
     pub async fn sync(&mut self) -> Result<(), Error> {
+        if self.mem_mmr.size() == 0 {
+            return Ok(());
+        }
+
         // Write the nodes cached in the memory-resident MMR to the journal.
         for i in self.journal_size..self.mem_mmr.size() {
-            let node = self.mem_mmr.get_node(i).await?.unwrap();
+            let node = *self.mem_mmr.get_node_unchecked(i);
             self.journal.append(node).await?;
         }
         self.journal_size = self.mem_mmr.size();
         self.journal.sync().await?;
         assert_eq!(self.journal_size, self.journal.size().await?);
 
-        // Clear out old peaks, then write the latest peaks to metadata.
-        self.metadata.clear();
-        let peak_iterator = self.mem_mmr.peak_iterator();
-        for (peak_pos, _) in peak_iterator {
-            let digest = self.mem_mmr.get_node(peak_pos).await?.unwrap();
-            self.metadata
-                .put(U64::new(peak_pos), Bytes::copy_from_slice(digest.as_ref()));
+        // Recompute pinned nodes since we'll need to repopulate the cache after it is cleared by
+        // pruning the mem_mmr.
+        let mut pinned_nodes = HashMap::new();
+        let required_positions = Proof::<H>::nodes_to_pin(self.mem_mmr.size(), self.pruned_to_pos);
+        for pos in required_positions.into_iter() {
+            let digest = self.mem_mmr.get_node_unchecked(pos);
+            pinned_nodes.insert(pos, *digest);
         }
-        self.metadata.sync().await.map_err(Error::MetadataError)?;
 
-        // Keep memory usage in check by pruning old nodes from the memory-resident MMR after
-        // they've been written to disk.
+        // Now that the pinned node set has been recomputed, it's safe to prune the mem_mmr and
+        // reinstate them.
         self.mem_mmr.prune_all();
+        self.mem_mmr.add_pinned_nodes(pinned_nodes);
 
         Ok(())
     }
 
+    /// Prefix used for nodes in the metadata prefixed U8 key.
+    const NODE_PREFIX: u8 = 0;
+
+    /// Prefix used for the key storing the prune_to_pos position in the metadata.
+    const PRUNE_TO_POS_PREFIX: u8 = 1;
+
+    /// Compute and add required nodes for the given pruning point to the metadata, and sync it to
+    /// disk. Return the computed set of required nodes.
+    async fn sync_metadata(&mut self, prune_to_pos: u64) -> Result<HashMap<u64, H::Digest>, Error> {
```

Comment:
nit: I think we should use a different word than "sync" here. Maybe `update` or `refresh`

Reply:
renamed to "update_metadata"